packages/timeline-state-resolver/src/service/stateHandler.ts

```diff
@@ -0,0 +1,220 @@
+import { FinishedTrace, startTrace, endTrace } from '../lib'
+import { Mappings, Timeline, TSRTimelineContent } from 'timeline-state-resolver-types'
+import { BaseDeviceAPI, CommandWithContext } from './device'
+import { Measurement, StateChangeReport } from './measure'
+
+interface StateChange<DeviceState, Command extends CommandWithContext> {
+	commands?: Command[]
+	state: Timeline.TimelineState<TSRTimelineContent>
+	deviceState: DeviceState
+	mappings: Mappings
+
+	measurement?: Measurement
+}
+interface ExecutedStateChange<DeviceState, Command extends CommandWithContext>
+	extends StateChange<DeviceState, Command> {
+	commands: Command[]
+}
+
+export class StateHandler<DeviceState, Command extends CommandWithContext> {
+	private stateQueue: StateChange<DeviceState, Command>[] = []
+	private currentState: ExecutedStateChange<DeviceState, Command> | undefined = undefined
+	private _executingStateChange = false
+
+	private clock: NodeJS.Timeout
+
+	private convertTimelineStateToDeviceState: BaseDeviceAPI<DeviceState, Command>['convertTimelineStateToDeviceState']
+	private diffDeviceStates: BaseDeviceAPI<DeviceState, Command>['diffStates']
+	private executeCommand: BaseDeviceAPI<DeviceState, Command>['sendCommand']
+
+	private logger: StateHandlerContext['logger']
+
+	constructor(
+		private context: StateHandlerContext,
+		private config: StateHandlerConfig,
+		device: BaseDeviceAPI<DeviceState, Command>
+	) {
+		this.logger = context.logger
+
+		this.convertTimelineStateToDeviceState = (s, m) => device.convertTimelineStateToDeviceState(s, m)
+		this.diffDeviceStates = (o, n, m) => device.diffStates(o, n, m)
+		this.executeCommand = async (c) => device.sendCommand(c)
+
+		this.clock = setInterval(() => {
+			// main clock to check if next state needs to be sent out
+			const nextStates = this.stateQueue.filter((s) => s.state.time < Date.now() + 20) // any states to be executed before the next tick
+
```

Comment:
should this be using `Date.now()` or the `getCurrentTime()` method being passed around?

```suggestion
this.stateQueue.forEach((nextState) => {
	const nextTime = Math.max(0, nextState.state.time - Date.now())
	if (nextTime > 20) return // Ignore any states for a future tick

```
This avoids needing to do a similar calculate twice, and reduces the amount of array iterations. It also clarifies that `nextTime` can only ever be 0-20